spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/DefaultPollableMessageSource.java

```diff
@@ -0,0 +1,253 @@
+	}
+
+	public void addInterceptor(ChannelInterceptor interceptor) {
+		this.interceptors.add(interceptor);
+	}
+
+	public void addInterceptor(int index, ChannelInterceptor interceptor) {
+		this.interceptors.add(index, interceptor);
+	}
+
+	@Override
+	public boolean isRunning() {
+		return this.running;
+	}
+
+	@Override
+	public void start() {
+		if (!this.running && this.source instanceof Lifecycle) {
+			((Lifecycle) this.source).start();
+		}
+		this.running = true;
+	}
+
+	@Override
+	public void stop() {
+		if (this.running && this.source instanceof Lifecycle) {
+			((Lifecycle) this.source).stop();
+		}
+		this.running = false;
+	}
+
+	/**
+	 * If there's a retry template, it will set the attributes holder via the listener. If
+	 * there's no retry template, but there's an error channel, we create a new attributes
+	 * holder here. If an attributes holder exists (by either method), we set the
+	 * attributes for use by the {@link ErrorMessageStrategy}.
+	 * @param message the Spring Messaging message to use.
+	 */
+	private void setAttributesIfNecessary(Message<?> message) {
+		boolean needHolder = this.errorChannel != null && this.retryTemplate == null;
+		boolean needAttributes = needHolder || this.retryTemplate != null;
+		if (needHolder) {
+			attributesHolder.set(ErrorMessageUtils.getAttributeAccessor(null, null));
+		}
+		if (needAttributes) {
+			AttributeAccessor attributes = attributesHolder.get();
+			if (attributes != null) {
+				attributes.setAttribute(ErrorMessageUtils.INPUT_MESSAGE_CONTEXT_KEY, message);
+				if (this.attributesProvider != null) {
+					this.attributesProvider.accept(attributes, message);
+				}
+			}
+		}
+	}
+
+	@Override
+	public boolean poll(MessageHandler handler) {
+		Message<?> message = this.source.receive();
+		if (message == null) {
+			return false;
+		}
+		AcknowledgmentCallback ackCallback = StaticMessageHeaderAccessor
+				.getAcknowledgmentCallback(message);
+		try {
+			if (this.retryTemplate == null && this.errorChannel == null) {
+				setAttributesIfNecessary(message);
+				doHandleMessage(handler, message);
+			}
+			else if (this.retryTemplate == null) {
+				try {
+					setAttributesIfNecessary(message);
+					doHandleMessage(handler, message);
+				}
+				catch (Exception e) {
+					this.messagingTemplate.send(this.errorMessageStrategy.buildErrorMessage(e, null));
+					return false;
+				}
+			}
+			else {
+				this.retryTemplate.execute(context -> {
+					setAttributesIfNecessary(message);
+					doHandleMessage(handler, message);
+					return null;
+				}, this.recoveryCallback);
+			}
+			AckUtils.autoAck(ackCallback);
+			return true;
+		}
+		catch (Exception e) {
+			AckUtils.autoNack(ackCallback);
+			throw (MessageHandlingException) e;
+		}
+	}
+
+	private void doHandleMessage(MessageHandler handler, Message<?> message) {
+		try {
+			handler.handleMessage(message);
+		}
+		catch (Exception e) {
+			throw new MessageHandlingException(message, e);
+		}
+	}
```

Comment:
Do we actually need `try/catch` here? It already throws ME by definition and I am assuming the `handler` in this case is always AMH, so it is handled there. 
Not sure, hence the question

Reply:
Need to add the `failedMessage` to the MHE - otherwise the `ErrorMessage` won't have a reference and we can't publish it to a DLQ with diagnostic headers.